daemon/cmd/fqdn.go

```diff
@@ -576,6 +576,39 @@ func (d *Daemon) notifyOnDNSMsg(lookupTime time.Time, ep *endpoint.Endpoint, epI
 	record.Log()
 
 	if msg.Response && msg.Rcode == dns.RcodeSuccess && len(responseIPs) > 0 {
+		// Create a critical section especially for when multiple DNS requests
+		// are in-flight for the same name (i.e. cilium.io).
+		//
+		// It is possible for the following race to occurr:
+		//
+		//              G1                               G2
+		//
+		// T0 --> NotifyOnDNSMsg()               NotifyOnDNSMsg()            <-- T0
+		//
+		// T1 --> UpdateGenerateDNS()            UpdateGenerateDNS()         <-- T1
+		//
+		//   T2 --> mutex.Lock()                 +---------------------------+
+		//                                       |No identities need updating|
+		//   T3 --> mutex.Unlock()               +---------------------------+
+		//
+		// T4 --> UpsertGeneratedIdentities()    UpsertGeneratedIdentities() <-- T4
+		//
+		//   T5 --> Upsert()                     DNS released back to app    <-- T5
+		//                                               |
+		// T6 --> DNS released back to app               |
+		//              |                                |
+		//              |                                |
+		//              v                                v
+		//       Traffic flows fine                Leads to policy drop
+		//
+		// Note how G2 releases the DNS msg back to the app at T5 because
+		// UpdateGenerateDNS() was a no-op. G1 had executed UpdateGenerateDNS()
+		// first at T1 and performed the necessary identitiy allocation. Due to
+		// G1 performing all the work first, G2 executes T4 as a no-op and
+		// releases the msg back to the app at T5 before G1 can.
+		d.notifyOnDNSMsgMu.Lock()
+		defer d.notifyOnDNSMsgMu.Unlock()
```

Comment:
This is serializing *all* DNS requests, is that necessary to fix the issue for DNS requests for the same name? Do you think that it is likely that this request handling is already serialized somewhat by the code patterns or will this be a performance regression? (For background, we use this fancy framework that splits DNS handling out into a bunch of independent goroutines so it seems a bit odd to then force the expensive part of handling the DNS responses to now be 100% serialized)

Reply:
no it's not necessary, i think we just did the simplest thing to see if it would fix the issue. we also talked about a fixed size list of mutexes and hash dns names to map to mutexes or something like that.

chris is doing perf testing right now 🧑‍🔬

Reply:
Fixed size list of mutexes implemented in last commit.